File: 프로그래머스/월간 코드 챌린지 시즌2/괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

int isRight(string s){
    stack<char> st;
    for (int i = 0; i < s.length(); i++) {
        if (st.empty() || s[i] == '(' || s[i] == '[' || s[i] == '{') {
            if (s[i] == '(') st.push(')');
            else if (s[i] == '[') st.push(']');
            else st.push('}');
        }
        else if (st.top() == s[i]) st.pop();
        else return 0;
    }

    return st.empty() ? 1 : 0;
}

int solution(string s) {
    int answer = 0;
    for(int i=0; i<s.length(); i++){
        // s[i]가 닫는 괄호면 건너뛰기
        if(s[i]==')' || s[i]==']' || s[i]=='}')
            continue;
        // 아니면 올바른 괄호 문자열인지 체크
        else answer += isRight(s.substr(i)+s.substr(0,i));
    }
    
    return answer;
}

File: 프로그래머스/월간 코드 챌린지 시즌2/괄호 회전하기/괄호 회전하기.py
def isRight(s):
    st = []
    for si in s:
        if len(st)==0 or si=='(' or si=='[' or si=='{':
            if si=='(': st.append(')')
            elif si=='[': st.append(']')
            else: st.append('}')
        elif st[-1] == si: st.pop(-1)
        else: return 0

    if len(st) == 0: return 1
    else: return 0

def solution(s):
    answer = 0
    for i in range(len(s)):
        if s[i]==')' or s[i]==']' or s[i]=='}': continue
        else: answer += isRight(s[i:]+s[:i])

    return answer
